Sources/TuistEnvKit/Installer/EnvInstaller.swift

```diff
@@ -0,0 +1,108 @@
+        switch self {
+        case .versionNotFound: return .abort
+        case .incompatibleSwiftVersion: return .abort
+        }
+    }
+
+    var description: String {
+        switch self {
+        case let .versionNotFound(version):
+            return "Version \(version) not found"
+        case let .incompatibleSwiftVersion(local, expected):
+            return "Found \(local) Swift version but expected \(expected)"
+        }
+    }
+
+    static func == (lhs: EnvInstallerError, rhs: EnvInstallerError) -> Bool {
+        switch (lhs, rhs) {
+        case let (.versionNotFound(lhsVersion), .versionNotFound(rhsVersion)):
+            return lhsVersion == rhsVersion
+        case let (.incompatibleSwiftVersion(lhsLocal, lhsExpected), .incompatibleSwiftVersion(rhsLocal, rhsExpected)):
+            return lhsLocal == rhsLocal && lhsExpected == rhsExpected
+        default:
+            return false
+        }
+    }
+}
+
+/// Class that manages the installation of Tuist versions.
+final class EnvInstaller: EnvInstalling {
+    // MARK: - Attributes
+
+    let buildCopier: BuildCopying
+    let versionsController: VersionsControlling
+
+    // MARK: - Init
+
+    init(buildCopier: BuildCopying = BuildCopier(),
+         versionsController: VersionsControlling = VersionsController())
+    {
+        self.buildCopier = buildCopier
+        self.versionsController = versionsController
+    }
+
+    // MARK: - Installing
+
+    func install(version: String) throws {
+        try withTemporaryDirectory { temporaryDirectory in
+            try install(version: version, temporaryDirectory: temporaryDirectory)
+        }
+    }
+
+    func install(version: String, temporaryDirectory: AbsolutePath) throws {
+        try installFromBundle(
+            bundleURL: URL(string: "https://github.com/tuist/tuist/releases/download/\(version)/tuistenv.zip")!,
+            version: version,
+            temporaryDirectory: temporaryDirectory
+        )
+    }
+
+    func installFromBundle(bundleURL: URL,
+                           version: String,
+                           temporaryDirectory: AbsolutePath) throws
+    {
+        let installationPath = try System.shared.which("tuist")
+
+        // Download bundle
+        logger.notice("Downloading TuistEnv version \(version)")
+        let downloadPath = temporaryDirectory.appending(component: Constants.envBundleName)
+        try System.shared.run(["/usr/bin/curl", "-LSs", "--output", downloadPath.pathString, bundleURL.absoluteString])
+
+        // Unzip
+        logger.notice("Installing...")
+        try System.shared.run(["/usr/bin/unzip", "-q", downloadPath.pathString, "tuistenv", "-d", temporaryDirectory.pathString])
+
+        // Copy
+        let cpArgs = ["cp", temporaryDirectory.appending(component: "tuistenv").pathString, installationPath]
+        do {
+            try System.shared.run(cpArgs)
+        } catch {
+            try System.shared.run(["sudo"] + cpArgs)
+        }
```

Comment:
Have you checked that it works fine even if the destination is the program that is doing the copy itself?